agent/xds/version_compat.go

```diff
@@ -0,0 +1,471 @@
+	stream ADSStream_v2
+	grpc.ServerStream
+}
+
+var _ ADSStream = (*adsStreamV3Shim)(nil)
+
+func (s *adsStreamV3Shim) Send(resp *envoy_discovery_v3.DiscoveryResponse) error {
+	respv2, err := convertDiscoveryResponseToV2(resp)
+	if err != nil {
+		return fmt.Errorf("Error converting a v3 DiscoveryResponse to v2: %w", err)
+	}
+
+	return s.stream.Send(respv2)
+}
+
+func (s *adsStreamV3Shim) Recv() (*envoy_discovery_v3.DiscoveryRequest, error) {
+	req, err := s.stream.Recv()
+	if err != nil {
+		return nil, err
+	}
+
+	reqv3, err := convertDiscoveryRequestToV3(req)
+	if err != nil {
+		return nil, fmt.Errorf("Error converting a v2 DiscoveryRequest to v3: %w", err)
+	}
+
+	return reqv3, nil
+}
+
+func convertDiscoveryRequestToV3(req *envoy_api_v2.DiscoveryRequest) (*envoy_discovery_v3.DiscoveryRequest, error) {
+	var pbuf proto.Buffer
+	if err := pbuf.Marshal(req); err != nil {
+		return nil, err
+	}
+
+	var reqV3 envoy_discovery_v3.DiscoveryRequest
+	if err := pbuf.Unmarshal(&reqV3); err != nil {
+		return nil, err
+	}
+
+	// only one field to munge
+	if err := convertTypeUrlsToV3(&reqV3.TypeUrl); err != nil {
+		return nil, err
+	}
+
+	return &reqV3, nil
+}
+
+func convertDiscoveryResponseToV2(resp *envoy_discovery_v3.DiscoveryResponse) (*envoy_api_v2.DiscoveryResponse, error) {
+	var pbuf proto.Buffer
+	if err := pbuf.Marshal(resp); err != nil {
+		return nil, err
+	}
+
+	var respV2 envoy_api_v2.DiscoveryResponse
+	if err := pbuf.Unmarshal(&respV2); err != nil {
+		return nil, err
+	}
+
+	if err := convertTypedConfigsToV2(&respV2); err != nil {
+		return nil, err
+	}
+
+	return &respV2, nil
+}
+
+// convertNetFilterToV2 is only used in tests.
+func convertNetFilterToV2(filter *envoy_listener_v3.Filter) (*envoy_listener_v2.Filter, error) {
+	var pbuf proto.Buffer
+	if err := pbuf.Marshal(filter); err != nil {
+		return nil, err
+	}
+
+	var filterV2 envoy_listener_v2.Filter
+	if err := pbuf.Unmarshal(&filterV2); err != nil {
+		return nil, err
+	}
+
+	if err := convertTypedConfigsToV2(&filterV2); err != nil {
+		return nil, err
+	}
+
+	return &filterV2, nil
+}
+
+// convertHttpFilterToV2 is only used in tests.
+func convertHttpFilterToV2(filter *envoy_http_v3.HttpFilter) (*envoy_http_v2.HttpFilter, error) {
+	var pbuf proto.Buffer
+	if err := pbuf.Marshal(filter); err != nil {
+		return nil, err
+	}
+
+	var filterV2 envoy_http_v2.HttpFilter
+	if err := pbuf.Unmarshal(&filterV2); err != nil {
+		return nil, err
+	}
+
+	if err := convertTypedConfigsToV2(&filterV2); err != nil {
+		return nil, err
+	}
+
+	return &filterV2, nil
+}
+
+// Responses
+func convertTypedConfigsToV2(pb proto.Message) error {
+	switch x := pb.(type) {
+	case *envoy_api_v2.DiscoveryResponse:
+		if err := convertTypeUrlsToV2(&x.TypeUrl); err != nil {
+			return fmt.Errorf("%T: %w", x, err)
+		}
+		for _, res := range x.Resources {
+			if err := convertTypedConfigsToV2(res); err != nil {
+				return fmt.Errorf("%T: %w", x, err)
+			}
+		}
+		return nil
+	case *any.Any:
```

Comment:
Minor nit, it might make more sense to pull this case to the front, since it is a little special. Not a blocker.

Reply:
I'll leave it for now.  The order of the switches _roughly_ reads top-to-bottom the way they would be first used. So for instance this function is about converting responses so you enter with a `DiscoveryResponse`. The bulk of a `DiscoveryResponse` is a list of generic `any.Any` resources which is why that section comes second, and so on.